tests/common/gen-bats.bash

```diff
@@ -59,30 +60,34 @@ render_test() {
   local target="${5:-}"
   local args="${6:-}"
 
-  if [[ -z "${clusters#[]}" ]]; then
-    log_fatal "render test: missing clusters"
-  elif [[ -z "${namespaces#[]}" ]]; then
-    log_fatal "render test: missing namespaces"
-  elif [[ -z "${function}" ]]; then
+  if [[ -z "${function}" ]]; then
     log_fatal "render test: missing function"
   elif [[ -z "${target}" ]]; then
     log_fatal "render test: missing target"
   fi
 
+  local identifier
+
+  identifier="${name} - $(sed -e 's/_/ /g' -e 's/.*\.//' <<< "${function}")"
+
   readarray -t clusters_arr <<< "$(jq -c '.[]' <<< "${clusters}")"
   for cluster in "${clusters_arr[@]}"; do
     readarray -t namespaces_arr <<< "$(jq -c '.[]' <<< "${namespaces}")"
     for namespace in "${namespaces_arr[@]}"; do
       echo ""
       # TODO: Set with flag
       if [ "${foreach:-[]}" != "[]" ]; then
-        echo "@test \"${name} - ${function//_/ } - ${cluster//\"/} / ${namespace//\"/} / ${target} - $(jq -r .[0] <<< "${args}")\" {"
+        echo "@test \"${identifier} - ${cluster//\"/} / ${namespace//\"/} / ${target} - $(jq -r .[0] <<< "${args}")\" {"
       else
-        echo "@test \"${name} - ${function//_/ } - ${cluster//\"/} / ${namespace//\"/} / ${target}\" {"
+        echo "@test \"${identifier} - ${cluster//\"/} / ${namespace//\"/} / ${target}\" {"
       fi
       render_conditions "${cluster}" "${conditions}"
-      echo "  with_kubeconfig ${cluster}"
-      echo "  with_namespace ${namespace}"
+      if [[ -n "${clusters#[]}" ]]; then
```

Comment:
Not sure I understand the reason for this. The way I interpret it is that you apply a "match nothing" pattern on the first element of the array which won't delete anything from it. What's different from just `-n "${clusters}"`?

Or in this case, couldn't we even assume that it never will be empty since we are looping over the array already?

Reply:
I didn't even realise how strange I did this. It works because an empty one will be `[]`, and if we have that then we don't render the `with`.

Reply:
Are you sure you're not thing of `${foo:-[]}`? That would result in `[]` if it's empty. `${foo#[]}` deletes with pattern matching AFAIK.

Reply:
I am bad at explaining today :sweat_smile: 

If `clusters` or `namespaces` is unset in input file then the variable will be set to `[]` (an empty array) and that is the condition I mean. `${var#[]}` does indeed remove the leading `[]` when it is expanded, and thus if `clusters` or `namespaces` with this expand to nothing then I consider it empty.

I should really just check `cluster` and `namespace` if they are empty. :grin:

Reply:
Now it all makes sense! :)